### PR TITLE
components: Limit the recursive size of types in Wasmtime

### LINKED ISSUE #4814: Limit type-based recursion in the component model
Currently the component model has no limit to type-bsaed recursion other than the "width" of a type and the number of types the type section is allowed to have. The fuzzers have subsequently studiously been finding all the issues with that in the implementation of the component model within Wasmtime. Wasmtime has a number of methods right now which recursively process a type which can lead to stack overflow for deeply nested types such as:

* the `Arbitrary for Type` implementation
* The `{Lower,Lift} for Val` implementations
* The compilation of adapter trampolines

In each of these cases it's not 100% clear if there's an easy way to switch stack-based recursion to heap-based recursion. So far switching to heap-based recursion or alternative methods has been the go-to for fixing fuzz-bugs like this (https://github.com/bytecodealliance/wasmtime/pull/4657 https://github.com/bytecodealliance/wasmtime/pull/4658 https://github.com/bytecodealliance/wasmtime/pull/4708 https://github.com/bytecodealliance/wasmtime/pull/4694) but this seems over time like it's starting to be untenable.

To solve all these issues the conclusion that @fitzgen @lukewagner and I reached awhile back was to simply limit recursion of type hierarchies in the component model entirely. This is a big hammer to use but would definitely fix all of the related issues here, in addition to the recent crop of fuzz-bugs which are getting harder and harder to fix via alternative means. This also seems to mirror what the GC proposal for wasm is doing where limits are placed on the type recursion there as well.

One option to implement this would be an error from `wasmparser` during validation but I think for now it may be best represented as an error from Wasmtime itself due to that being where all the fuzz bugs are. If we feel it's necessary this could also be moved to `wasmparser` instead. 

One bit of trickiness here is that we'll need to update `Arbitrary for Type` to have some limited recursion, and I'm not sure how to do that with `#[derive]` without a thread-local based solution. This may be a case where it's better to have a custom method taking `Unstructured` with a `depth` where the `depth` filters out some possible types to generate.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is aimed at fixing #4814 by placing a hard limit on the
maximal recursive depth a type may have in the component model. The
component model theoretically allows for infinite recursion but many
various types of operations within the component model are naturally
written as recursion over the structure of a type which can lead to
stack overflow with deeply recursive types. Some examples of recursive
operations are:

* Lifting and lowering a type - currently the recursion here is modeled
  in Rust directly with `#[derive]` implementations as well as the
  implementations for the `Val` type.

* Compilation of adapter trampolines which iterates over the type
  structure recursively.

* Historically many various calculations like the size of a type, the
  flattened representation of a type, etc, were all done recursively.
  Many of these are more efficiently done via other means but it was
  still natural to implement these recursively initially.

By placing a hard limit on type recursion Wasmtime won't be able to load
some otherwise-valid modules. The hope, though, is that no human-written
program is likely to ever reach this limit. This limit can be revised
and/or the locations with recursion revised if it's ever reached.

The implementation of this feature is done by generalizing the current
flattened-representation calculation which now keeps track of a type's
depth and size. The size calculation isn't used just yet but I plan to
use it in fixing #4816 and it was natural enough to write here as well.
The depth is checked after a type is translated and if it exceeds the
maximum then an error is returned.

Additionally the `Arbitrary for Type` implementation was updated to
prevent generation of a type that's too-recursive.

Closes #4814

### CODE REVIEW

**Comment by alexcrichton** [comment]
> I plan to use it in fixing https://github.com/bytecodealliance/wasmtime/issues/4816

Turns out this was a lie and I didn't use it when implementing https://github.com/bytecodealliance/wasmtime/pull/4827. I'll remove this code tomorrow.

File: crates/environ/src/fact/trampoline.rs [hunk 1]
```diff
@@ -78,52 +78,69 @@ pub(super) fn compile(module: &mut Module<'_>, adapter: &AdapterData) {
...
-/// `src_opts` and `dst_opts` specified as well.
-pub(super) fn compile_translate_mem(
-    module: &mut Module<'_>,
-    result: FunctionId,
-    src: InterfaceType,
-    src_opts: &Options,
-    dst: InterfaceType,
-    dst_opts: &Options,
-) {
+/// TODO
```
Comment (fitzgen) [inline crates/environ/src/fact/trampoline.rs:86]: todo

File: crates/misc/component-fuzz-util/src/lib.rs [hunk 2]
```diff
@@ -19,6 +19,9 @@ const MAX_FLAT_PARAMS: usize = 16;
 const MAX_FLAT_RESULTS: usize = 1;
 const MAX_ARITY: u32 = 5;
 
+// Wasmtime allows up to 100 type depth so limit this to just under that.
+const MAX_TYPE_DEPTH: u32 = 90;
```
Comment (fitzgen) [inline crates/misc/component-fuzz-util/src/lib.rs:23]: Why not 99?

**Review by fitzgen — APPROVED** [review]
Nice!

File: crates/environ/src/fact/trampoline.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/environ/src/fact/trampoline.rs]: Oops this was actually intended for a separate PR so I'm going to remove this

File: crates/misc/component-fuzz-util/src/lib.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/misc/component-fuzz-util/src/lib.rs]: Originally I didn't really want to deal with off-by-one bugs with these sorts of arbitrary limits, but I suppose there's no harm in fixing them if they come up.

**Pull request merged**